genie-agent/src/main/java/com/netflix/genie/agent/spring/autoconfigure/AgentAutoConfiguration.java

```diff
@@ -90,46 +92,93 @@ public FileLockFactory fileLockFactory() {
+    public AsyncTaskExecutor sharedAgentTaskExecutor(final AgentProperties agentProperties) {
         final ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
         executor.setCorePoolSize(5);
         executor.setThreadNamePrefix("agent-task-executor-");
+        executor.setWaitForTasksToCompleteOnShutdown(true);
+        executor.setAwaitTerminationSeconds(
+            (int) agentProperties.getShutdown().getInternalExecutorsLeeway().getSeconds()
+        );
         return executor;
     }
 
     /**
      * Provide a lazy {@link TaskScheduler} to be used by the Agent process if one isn't already defined.
      *
+     * @param agentProperties the agent properties
      * @return A {@link ThreadPoolTaskScheduler} instance
      */
     @Bean
     @Lazy
     @ConditionalOnMissingBean(name = "sharedAgentTaskScheduler")
-    public TaskScheduler sharedAgentTaskScheduler() {
+    public TaskScheduler sharedAgentTaskScheduler(final AgentProperties agentProperties) {
         final ThreadPoolTaskScheduler scheduler = new ThreadPoolTaskScheduler();
         scheduler.setPoolSize(5); // Big enough?
         scheduler.setThreadNamePrefix("agent-task-scheduler-");
+        scheduler.setWaitForTasksToCompleteOnShutdown(true);
+        scheduler.setAwaitTerminationSeconds(
+            (int) agentProperties.getShutdown().getInternalSchedulersLeeway().getSeconds()
+        );
         return scheduler;
     }
 
     /**
      * Provide a lazy {@link TaskScheduler} bean for use by the heart beat service is none has already been
      * defined in the context.
      *
+     * @param agentProperties the agent properties
      * @return A {@link TaskScheduler} that the heart beat service should use
      */
     @Bean
     @Lazy
     @ConditionalOnMissingBean(name = "heartBeatServiceTaskScheduler")
-    public TaskScheduler heartBeatServiceTaskScheduler() {
+    public TaskScheduler heartBeatServiceTaskScheduler(final AgentProperties agentProperties) {
         final ThreadPoolTaskScheduler taskScheduler = new ThreadPoolTaskScheduler();
         taskScheduler.setPoolSize(1);
         taskScheduler.initialize();
+        taskScheduler.setWaitForTasksToCompleteOnShutdown(true);
+        taskScheduler.setAwaitTerminationSeconds(
+            (int) agentProperties.getShutdown().getInternalSchedulersLeeway().getSeconds()
+        );
         return taskScheduler;
     }
+
+    /**
+     * Customizer for Spring's task executor.
+     *
+     * @param agentProperties the agent properties
+     * @return a customizer for the task executor
+     */
+    @Bean
+    TaskExecutorCustomizer taskExecutorCustomizer(final AgentProperties agentProperties) {
+        return taskExecutor -> {
+            taskExecutor.setWaitForTasksToCompleteOnShutdown(true);
+            taskExecutor.setAwaitTerminationSeconds(
+                (int) agentProperties.getShutdown().getSystemExecutorLeeway().getSeconds()
```

Comment:
You may want to see if you can set a `@min` in your property class for `Duration` objects or some sort of validation because if the person sets like `400ms` as the value (totally hypothetical) I think this will round it down to 0 seconds

Reply:
@tgianos added a commit in this PR that adds validation to all agent properties.